### PR TITLE
WIP: Add perf-analyzer-no-security integ tests to staging-build-tar workflow

### PR DESCRIPTION
**Pull request by camerski**
### Issue
https://github.com/opendistro-for-elasticsearch/opendistro-infra/issues/236
https://github.com/opendistro-for-elasticsearch/opendistro-build/issues/339

### Description of changes
Add perf-analyzer-no-security integ tests to staging-build-tar workflow.

### Test Results
Manually tested the workflow. The integ tests start, but fail right now because the tests need to enable a feature in the test cluster.

https://github.com/opendistro-for-elasticsearch/opendistro-build/actions/runs/217498888

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/staging-build-tar.yml [hunk 1]
```diff
@@ -459,3 +459,50 @@ jobs:
           working-directory: kibana/plugins/anomaly-detection-kibana-plugin
           command: yarn cy:run-with-security --config baseurl=http://localhost:5601
     
+  Test-Perf-Analyzer-No-Security:
+    runs-on: ubuntu-18.04
+    name: Test-Perf-Analyzer-No-Security
+    strategy:
+      matrix:
+        java: [12]
```
Comment (peterzhuamazon) [inline .github/workflows/staging-build-tar.yml:467]: Please add a comment in the yml on java 12, this should be a temp measure and we should eventually switch to 14 just like other plugins.

File: .github/workflows/staging-build-tar.yml [hunk 2]
```diff
@@ -459,3 +459,50 @@ jobs:
           working-directory: kibana/plugins/anomaly-detection-kibana-plugin
           command: yarn cy:run-with-security --config baseurl=http://localhost:5601
     
+  Test-Perf-Analyzer-No-Security:
+    runs-on: ubuntu-18.04
+    name: Test-Perf-Analyzer-No-Security
```
Comment (peterzhuamazon) [inline .github/workflows/staging-build-tar.yml:464]: This is not following the order we are doing in all the tests.
This test should be after "Test-AD-NoSec".

Also, the naming convention is not on par, you should use "Test-PA-NoSec".

File: .github/workflows/staging-build-tar.yml [hunk 3]
```diff
@@ -459,3 +459,50 @@ jobs:
           working-directory: kibana/plugins/anomaly-detection-kibana-plugin
           command: yarn cy:run-with-security --config baseurl=http://localhost:5601
     
+  Test-Perf-Analyzer-No-Security:
+    runs-on: ubuntu-18.04
```
Comment (peterzhuamazon) [inline .github/workflows/staging-build-tar.yml:463]: [ubuntu-18.04], follow the pattern.

File: .github/workflows/staging-build-tar.yml [hunk 4]
```diff
@@ -459,3 +459,50 @@ jobs:
...
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-east-1
+      - name: Checkout opendistro-build
+        uses: actions/checkout@v1
+      - name: Checkout perf analyzer RCA package
+        uses: actions/checkout@v1
+        with:
+          repository: opendistro-for-elasticsearch/performance-analyzer-rca
+          # TODO: Change to a tag once new tags have been published
+          ref: master
```
Comment (peterzhuamazon) [inline .github/workflows/staging-build-tar.yml:486]: Not only you directly checkout master in the yml, but also you did not retrieve the tag to checkout from there.

- name: Retrieve plugin tags
        run:  echo ::set-env name=p_tag_pa::$(.github/scripts/plugin_tag.sh opendistro-for-elasticsearch/performance-analyzer)

- name: Retrieve plugin tags
        run:  echo ::set-env name=p_tag_pa_rca::$(.github/scripts/plugin_tag.sh opendistro-for-elasticsearch/performance-analyzer=rca)

File: .github/workflows/staging-build-tar.yml [hunk 5]
```diff
@@ -459,3 +459,50 @@ jobs:
...
+      - name: Build RCA into MavenLocal
+        run: |
+          cd ${GITHUB_WORKSPACE}/../performance-analyzer-rca
+          ./gradlew publishToMavenLocal
+      - name: Update SHA
+        run: |
+          cd ${GITHUB_WORKSPACE}/../performance-analyzer
+          rm licenses/performanceanalyzer-rca-1.3.jar.sha1
+          ./gradlew updateShas
```
Comment (peterzhuamazon) [inline .github/workflows/staging-build-tar.yml:501]: I do not think build into mavenlocal and updateshas is necessary for our purpose here.

File: .github/workflows/staging-build-tar.yml [hunk 6]
```diff
@@ -459,3 +459,50 @@ jobs:
...
+      - name: Run integration tests
+        run: |
+          .github/scripts/setup_runners_service.sh zip --es-nosec
+          export PATH=$JAVA_HOME:$PATH
+          cd ${GITHUB_WORKSPACE}/../performance-analyzer
+          pwd
+          ./gradlew integTest -Dtests.enableIT=true -Dtests.useDockerCluster=false -Dtests.rest.cluster=localhost:9200 -Dtests.cluster=localhost:9300
```
Comment (peterzhuamazon) [inline .github/workflows/staging-build-tar.yml:508]: Sid has mentioned in the issue on these two specific curl to do after starting ES:


If an ES cluster is running with ODFE 1.9+ with PA 1.9+, the following commands will prep the cluster (enable PA and RCA):

    curl localhost:9200/_opendistro/_performanceanalyzer/cluster/config -H 'Content-Type: application/json' -d '{"enabled": true}'
    curl localhost:9200/_opendistro/_performanceanalyzer/rca/cluster/config -H 'Content-Type: application/json' -d '{"enabled": true}'

Comment (gaiksaya) [reply .github/workflows/staging-build-tar.yml:508]: Yeah add those lines here https://github.com/opendistro-for-elasticsearch/opendistro-build/pull/354/files#diff-dcd3e1c66f2aa852aabbbd3739826be3R507 
I am assuming this is feature you are talking about in **Test Results** section in order to pass the test

**Comment by peterzhuamazon** [comment]
If this PR is not in ready state for review, e.g. has many TODOs. Please add "WIP" in the title, and do not assign us for review until everything is ready. Thanks.

**Comment by camerski** [comment]
> should we enable the PA feature firstly? I see the " PA not enabled" error in the log

I tried that; the tests still fail. See https://github.com/opendistro-for-elasticsearch/opendistro-build/actions/runs/219069273/workflow

Sid's change should address this: https://github.com/opendistro-for-elasticsearch/performance-analyzer/pull/176/

File: .github/workflows/staging-build-tar.yml [hunk 7]
```diff
@@ -216,7 +216,58 @@ jobs:
...
+          ref: ${{env.pa_tag}}
+      - name: Build RCA into MavenLocal
+        run: |
+          cd ${GITHUB_WORKSPACE}/../performance-analyzer-rca
+          ./gradlew publishToMavenLocal
+      - name: Update SHA
+        run: |
+          cd ${GITHUB_WORKSPACE}/../performance-analyzer
+          rm licenses/performanceanalyzer-rca-1.3.jar.sha1
+          ./gradlew updateShas
```
[the remark this answers is not in the document]
Comment (camerski) [reply .github/workflows/staging-build-tar.yml:263]: Unfortunately this is still needed right now. RCA-1.3 is not published to Maven Central, so the :compileJava task fails.

There is a chicken-and-egg problem here - we can't publish RCA-1.3 publicly until the integ tests pass, and the tests won't pass until RCA-1.3 is published. I'm going to leave the local build in place to work around this problem.

**Comment by peterzhuamazon** [comment]
> > should we enable the PA feature firstly? I see the " PA not enabled" error in the log
> 
> I tried that; the tests still fail. See https://github.com/opendistro-for-elasticsearch/opendistro-build/actions/runs/219069273/workflow
> 
> Sid's change should address this: [opendistro-for-elasticsearch/performance-analyzer#176](https://github.com/opendistro-for-elasticsearch/performance-analyzer/pull/176)

Ping him and ask when he can get that PR merged, and create a tag for 1.10.0 then test again.

**Comment by camerski** [comment]
> > > should we enable the PA feature firstly? I see the " PA not enabled" error in the log
> > 
> > 
> > I tried that; the tests still fail. See https://github.com/opendistro-for-elasticsearch/opendistro-build/actions/runs/219069273/workflow
> > Sid's change should address this: [opendistro-for-elasticsearch/performance-analyzer#176](https://github.com/opendistro-for-elasticsearch/performance-analyzer/pull/176)
> 
> Ping him and ask when he can get that PR merged, and create a tag for 1.10.0 then test again.

Way ahead of you :)

[August 24, 2020, 9:20 AM] Skinner, Cameron: Hey Sid, what's the ETA for getting that change merged?
[August 24, 2020, 9:24 AM] Narayan, Sid: Let me ping people, I'll try to get it merged by EOD, hopefully sooner

**Comment by peterzhuamazon** [comment]
Talking to Sid about this in a meeting, and the outcomes are:
* We have to have RCA jars build fresh because:
  * gradle integTest framework requires a full code compilation to run test
  * RCA jar is a dependency for PA jar
  * Since RCA and PA sits in two different repo instead of one, during the integTest compileJAVA step, test will fail if it could not find a RCA jar presents in the same repo parent folder as PA (/tmp/pa, /tmp/rca structure required here)
  * The solution is to build RCA jars on the fly, push it to mavenlocal, so the PA integTest can find it as a dependency through mavenlocal, thus use it for full compilation during integTest compileJAVA step
* However, the integTest is only doing API calls through 9200 port, does not necessarily need the code in RCA jar to actually run the integTest
* This would be taken care of by PA team later
* They will also try to upgrade their gradle to the latest version in order to take JDK 14/15 later

**Comment by peterzhuamazon** [comment]
Close this as the integTest model changed in OpenSearch.

**Pull request closed without merging**